stan/math/opencl/matrix_cl.hpp

```diff
@@ -21,33 +22,41 @@
  */
 namespace stan {
 namespace math {
+
+// Dummy class to instantiate matrix_cl to enable for specific types.
+template <typename T, typename = void>
+class matrix_cl {};
+
```

Comment:
@andrjohns So we can actually get away with this. Was going through the C++ templates book section on template lookup rules / partial specializations.

Essentially, the compiler only goes with this default template if there is nothing else to go on. Then the below `matrix_cl` just becomes a standard template specialization that's preferred for anything arithmetic.


 I'm kind of tempted to get rid of the default constructor and have it throw and error like "This template class should never be constructed"